### PR TITLE
Iz gs preview modal

### PR DESCRIPTION
**Pull request by zarsky-broad**
~~Still needs styling pass, but initial review would be appreciated.~~

Now pretty. Not, like, final-pretty, but for-now-pretty.

### CODE REVIEW

File: config/dev.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
   "firecloudUrlRoot": "https://firecloud.dsde-dev.broadinstitute.org",
   "googleClientId": "500025638838-s2v23ar3spugtd5t2v1vgfa2sp7ppg0d.apps.googleusercontent.com",
   "leoUrlRoot": "https://leonardo.dsde-dev.broadinstitute.org",
+  "marthaUrlRoot": "https://us-central1-broad-dsde-prod.cloudfunctions.net/martha_v1",
```
Comment (panentheos) [inline config/dev.json:7]: Do we mean to be using prod here?

File: src/libs/ajax.js [hunk 2]
```diff
@@ -84,6 +89,20 @@ const fetchBuckets = (path, ...args) => fetchOk(`https://www.googleapis.com/${pa
 const nbName = name => encodeURIComponent(`notebooks/${name}.ipynb`)
 
 export const Buckets = {
+  getObject: async (bucket, object, namespace) => {
+    return await fetchOrchestration(`/api/storage/${bucket}/${object}`,
```
Comment (panentheos) [inline src/libs/ajax.js]: nit: `return await ...` is usually extraneous, you can just return the promise directly.

File: src/libs/data-utils.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import { spinner } from 'src/components/icons'
+import Modal from 'src/components/Modal'
+import { TextCell } from 'src/components/table'
+import { Buckets, Martha } from 'src/libs/ajax'
+import * as Utils from 'src/libs/utils'
+import { Component } from 'src/libs/wrapped-components'
+
+
+const styles = {
+  cell: children => div({ style: { marginBottom: '0.5rem' } }, children),
```
Comment (panentheos) [inline src/libs/data-utils.js]: This seems misplaced to me because it's effectively a component, not a style. I think the abstraction is fine, but perhaps it should be a standalone function for clarity.

File: src/libs/data-utils.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  label: text => div({}, text),
+  data: children => div({}, children)
```
Comment (panentheos) [inline src/libs/data-utils.js]: `label` and `data` look unused

File: src/libs/data-utils.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+const isFilePreviewable = name => /\.txt$/.test(name) || /\.[ct]sv$/.test(name) || /\.log/.test(name)
+const formatPriceInDollars = price => price < 0.01 ? '< $0.01' : `$${parseFloat(price).toFixed(2)}`
```
Comment (panentheos) [inline src/libs/data-utils.js]: TOL: I think there are built-in currency formatters in the browser. Not sure if they would actually add any value here, just mentioning.

File: src/libs/data-utils.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  async getMetadata() {
+    const { googleProject } = this.props
+    const { uri } = this.state
+    const [bucket, object] = parseUri(uri)
+
+    this.setState({
+      metadata: await Buckets.getObject(bucket, object, googleProject),
+      preview: isFilePreviewable(uri) ? await Buckets.getObjectPreview(bucket, object, googleProject) : undefined
```
Comment (panentheos) [inline src/libs/data-utils.js]: If these two calls are independent, they could be parallelized.

File: src/libs/data-utils.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    super(props)
+
+    const { uri } = props
+
+    this.state = { uri: _.startsWith('gs://', uri) ? uri : undefined }
+  }
+
+  async getMetadata() {
+    const { googleProject } = this.props
+    const { uri } = this.state
```
Comment (panentheos) [inline src/libs/data-utils.js]: In the case of the uri being resolved by Martha, I don't think the `setState` is guaranteed to be visible by the time we get here. It might be safer to pass the value directly, and/or combine these methods.

File: src/libs/data-utils.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+                    h(Collapse, { title: 'More Information', defaultHidden: true }, [
+                      styles.cell([`Created: ${Utils.makePrettyDate(metadata.timeCreated)}`]),
+                      styles.cell([`Updated: ${Utils.makePrettyDate(metadata.updated)}`]),
+                      styles.cell([`md5: ${metadata.md5Hash}`])
+                    ])
+                  ]
+                )
+              ]
+            }
+          ], () => ['Resolving DOS uri...', spinner()])
```
Comment (panentheos) [inline src/libs/data-utils.js]: TOL: Would it be simpler to unify the spinners here, or do these operations take long enough that it's helpful to give intermediate feedback/status?

File: src/libs/data-utils.js [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                    ] : ['File can\'t be previewed.']),
+                    styles.cell([`File size: ${filesize(metadata.size)}`]),
+                    styles.cell([
+                      link({
+                        target: 'blank',
+                        href: `https://accounts.google.com/AccountChooser?continue=https://console.cloud.google.com/storage/browser/${parseUri(
+                          uri)[0]}`
+                      }, ['View this file in the Google Cloud Storage Browser'])
+                    ]),
+                    styles.cell([buttonPrimary({}, [`Download for ${formatPriceInDollars(metadata.estimatedCostUSD)}`])]),
```
Comment (panentheos) [inline src/libs/data-utils.js]: It looks like this button does nothing.

File: config/dev.json (hunk 1 above)
Comment (zarsky-broad) [reply config/dev.json:7]: That's what the FC configs say for dev, I copied them.

**Review by MattPutnam — APPROVED** [review]
👍

**Review by panentheos — APPROVED** [review]
Looks good.
